### PR TITLE
NDEF framework

### PR DESCRIPTION
**Pull request by pan-**
### Description

NDEF framework inplementation.

### Pull request type

    [ ] Fix
    [ ] Refactor
    [ ] New target
    [X] Feature
    [ ] Breaking change

### CODE REVIEW

File: features/nfc/nfc/common/Mime.h [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    );
+
+    /**
+     * Return the mime type.
+     * @return The mime type.
+     */
+    Span<const uint8_t> get_mime_type() const;
+
+    /**
+     * Return the content of thr mime object.
```
Comment (donatieng) [inline features/nfc/nfc/common/Mime.h]: "thr"

File: features/nfc/nfc/common/Mime.h [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    Mime();
+
+    /**
+     * Construct a mime object from its type and content
+     *
+     * @param mime_type The mime type of the object
+     * @param content The content of the object.
+     */
+    Mime(
+        const Span<const uint8_t> &mime_type,
```
Comment (donatieng) [inline features/nfc/nfc/common/Mime.h]: A Mime type is a string, could we have a helper accepting a `const char*`?

File: features/nfc/nfc/common/Text.h [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+    /**
+     * Construct a text element from a data buffer and an encoding.
+     * @param text_encoding The encoding of the text.
+     * @param language_code The string of the language code.
+     * @param text The text buffer.
+     */
+    Text(
+        encoding_t text_encoding,
+        const Span<const uint8_t> &language_code,
+        const Span<const uint8_t> &text
```
Comment (donatieng) [inline features/nfc/nfc/common/Text.h]: Same comment as for Mime type

File: features/nfc/nfc/common/URI.h [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+        POP = 0x14, /// pop:
+        SIP = 0x15, /// sip:
+        SIPS = 0x16, /// sips:
+        TFTP = 0x17, /// tftp:
+        BTSPP = 0x18, /// btspp://
+        BTL2CAP = 0x19, /// btl2cap://
+        BTGOEP = 0x1A, /// btgoep://
+        TCPOBEX = 0x1B, /// tcpobex://
+        IRDAOBEX = 0x1C, /// irdaobex://
+        FILE = 0x1D, /// file://
```
Comment (donatieng) [inline features/nfc/nfc/common/URI.h]: This will clash if `stdio.h` is included

File: features/nfc/nfc/common/Mime.h [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+#define NFC_COMMON_MIME_H_
+
+#include "platform/Span.h"
+#include <nfc/ndef/RecordParser.h>
+#include <nfc/ndef/MessageBuilder.h>
+#include <stdint.h>
+
+namespace mbed {
+namespace nfc {
+namespace common {
```
Comment (donatieng) [inline features/nfc/nfc/common/Mime.h]: This is very NDEF-related, shouldn't it be in the NDEF namespace?

File: features/nfc/nfc/common/URI.h [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+     */
+    URI& operator=(const URI &to_copy);
+
+    /**
+     * Replace the value of the URI.
+     * @param id The ID of the URI
+     * @param uri_field A buffer containing the value of the URI field.
+     */
+    void set_uri(
+        uri_identifier_code_t id,
```
Comment (donatieng) [inline features/nfc/nfc/common/URI.h]: If not too much work, could we also have a method that converts a full URI (including NDEF prefix) to prefix/rest of URI, and vice-versa?

**Review by donatieng — CHANGES_REQUESTED** [review]
Thanks @pan-, some comments/notes

File: features/nfc/nfc/common/URI.h (hunk 4 above)
Comment (pan-) [reply features/nfc/nfc/common/URI.h]: It shouldn't, it is not the same scope: FILE is a typedef in the global scope and this is an enum inside the URI class; it is only accessible with URI::FILE.

File: features/nfc/nfc/common/Mime.h (hunk 5 above)
Comment (pan-) [reply features/nfc/nfc/common/Mime.h]: Good point, should it be `ndef::common`  or just inside `ndef` ?

File: features/nfc/nfc/common/URI.h (hunk 4 above)
Comment (donatieng) [reply features/nfc/nfc/common/URI.h]: Good point :)

**Comment by pan-** [comment]
@donatieng I've addressed your comments except the one about URI. I will add these conversion functions  during the week end.

**Review by donatieng — APPROVED** [review]
OK thanks, I'll go ahead and merge this as I want this to be visible for review within the main PR

**Pull request merged**